working/0546-patterns/patterns-feature-specification.md

```diff
@@ -2038,16 +2069,38 @@ To match a pattern `p` against a value `v`:
 
     1.  Evaluate the right-hand constant expression to `c`.
 
-    2.  A `== c` pattern matches if `v == c` evaluates to true. *This takes into
-        account the built-in semantics that `null` is only equal to `null`.*
+    2.  If the operator is `==`:
+
+        1.  Let `r` be the result of `v == c`.
+
+        2.  If `r` is not a Boolean then throw a runtime error. *This can
+            happen if operator `==` on `v`'s type returns `dynamic`.*
+
+        3.  The pattern matches if `r` is true and fails otherwise. *This takes
+            into account the built-in semantics that `null` is only equal to
+            `null`.*
+
+    2.  Else if the operator is `!=`:
+
+        1.  Let `r` be the result of `v == c`.
+
+        2.  If `r` is not a Boolean then throw a runtime error. *This can
+            happen if operator `==` on `v`'s type returns `dynamic`.*
+
+        3.  The pattern matches if `r` is false and fails otherwise. *This takes
+            into account the built-in semantics that `null` is only equal to
+            `null`.*
+
+    3.  Else the operator is a comparison operator `op`:
 
-    3.  A `!= c` pattern matches if `v == e` evaluates to false. *This takes
-        into account the built-in semantics that `null` is not equal to anything
-        but `null`.*
+        1.  Let `r` be the result of calling `op` on `v` with argument `c`.
 
-    4.  For any other operator, the pattern matches if calling the operator
-        method of the same name on the matched value, with `c` as the argument
-        returns true.
+        2.  If `r` is not a Boolean then throw a runtime error. *This can happen
+            if the operator on `v`'s type returns `dynamic`.*
```

Comment:
> Is that pattern allowed? (I can argue either way.)

It's not. Under type checking, we say:

> 1.  Type-check each of `f`'s positional field subpatterns using the corresponding positional field type on `M` as the matched value type **or `Object?` if `M` is not a record type with the corresponding field.**

(Emphasis added.) Since we don't know the types of the fields, we type check each field as `Object?`. Since `Object?` doesn't define `<` (unless you've got some extension floating around that adds it), this is a compile error.

> If no, could I do `dynamic _ & < "a"` instead to force the dynamic invocation?

We still need to figure out promotion but... probably not?

> (Would it change if the switch expression had type `dynamic` instead of `Object?`?)

No, strangely. It will still use `Object?`. That's a bug. Rewrote the type checking to fix.